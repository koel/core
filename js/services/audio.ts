interface AudioService {
  context: AudioContext | null
  source: MediaElementAudioSourceNode | null
  element: HTMLMediaElement | null

  init (element: HTMLMediaElement): void
  getContext(): AudioContext
  getSource(): MediaElementAudioSourceNode
  getElement(): HTMLMediaElement
}

export const audio: AudioService = {
  context: null,
  source: null,
  element: null,

  init (element: HTMLMediaElement): void {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext () {
    return this.context!
  },

  getSource () {
    return this.source!
  },

  getElement () {
    return this.element!
  }
}
