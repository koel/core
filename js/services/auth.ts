import { ls } from '.'

export const auth = {
  storageKey: 'api-token',

  getToken (): string | null {
    return ls.get<string | null>(this.storageKey)
  },

  hasToken (): boolean {
    return Boolean(this.getToken())
  },

  setToken (token: string): void {
    ls.set(this.storageKey, token)
  },

  destroy (): void {
    ls.remove(this.storageKey)
  }
}
