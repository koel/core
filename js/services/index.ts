export * from './http'
export * from './info'
export * from './download'
export * from './ls'
export * from './playback'
export * from './youtube'
export * from './socket'
export * from './audio'
export * from './upload'
export * from './auth'
