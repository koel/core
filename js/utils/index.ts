export * from './event'
export * from './alerts'
export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
export * from './$'
export * from './app'
export * from './helpers'
export * from './file-reader'
