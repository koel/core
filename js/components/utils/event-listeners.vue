<script lang="ts">
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import Vue, { VNode } from 'vue'
import isMobile from 'ismobilejs'
import router from '@/router'
import { auth } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'
import { events } from '@/config'

export default Vue.extend({
  render: (h: Function): VNode => h(),
  created (): void {
    eventBus.on({
      [events.PLAYLIST_DELETE]: (playlist: Playlist): void => {
        const destroy = async (): Promise<void> => {
          await playlistStore.delete(playlist)
          alerts.success(`Deleted playlist &quot;${playlist.name}&quot;.`)
          router.go('home')
        }

        if (!playlist.songs.length) {
          destroy()
        } else {
          alerts.confirm(`Delete the playlist &quot;${playlist.name}&quot?`, destroy)
        }
      },

      /**
       * Log the current user out and reset the application state.
       */
      [events.LOG_OUT]: async (): Promise<void> => {
        await userStore.logout()
        auth.destroy()
        forceReloadWindow()
      },

      [events.KOEL_READY]: (): void => router.init(),

      /**
       * Hide the panel away if a main view is triggered on mobile.
       */
      [events.LOAD_MAIN_CONTENT]: (): void => {
        if (isMobile.phone) {
          preferenceStore.showExtraPanel = false
        }
      }
    })
  }
})
</script>
