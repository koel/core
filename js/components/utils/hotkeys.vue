<template>
  <global-events
    @keydown.space = "togglePlayback"
    @keydown.j = "playNext"
    @keydown.k = "playPrev"
    @keydown.f = "search"
    @keydown.l = "toggleLike"
    @keydown.mediaPrev = "playPrev"
    @keydown.mediaNext = "playNext"
    @keydown.mediaToggle = "togglePlayback"
  />
</template>

<script lang="ts">
import Vue from 'vue'
import GlobalEvents from 'vue-global-events'
import { $, event, noop } from '@/utils'
import { playback, socket } from '@/services'
import { queueStore, favoriteStore, songStore } from '@/stores'

let ipc: any, events: any

if (KOEL_ENV === 'app') {
  ipc = require('electron').ipcRenderer
  events = require('&/events').default
}

// Register our custom key codes
Vue.config.keyCodes = {
  a: 65,
  j: 74,
  k: 75,
  f: 70,
  l: 76,
  mediaNext: 176,
  mediaPrev: 177,
  mediaToggle: 179
}

/**
 * Listen to the global shortcuts (media keys).
 * Only works in the app.
 */
let listenToGlobalShortcuts = noop

if (KOEL_ENV === 'app') {
  listenToGlobalShortcuts = (): void => {
    const mediaFunctionMap = {
      MediaNextTrack: () => playback.playNext(),
      MediaPreviousTrack: () => playback.playPrev(),
      MediaStop: () => playback.stop(),
      MediaPlayPause: () => playback.toggle()
    } as { [propName: string]: Function }

    ipc.on(events.GLOBAL_SHORTCUT, (e: any, msg: string) => msg in mediaFunctionMap && mediaFunctionMap[msg]())
  }
}

export default Vue.extend({
  components: {
    GlobalEvents
  },

  methods: {
    /**
     * Toggle playback when user presses Space key.
     *
     * @param {Object} e The keydown event
     */
    togglePlayback: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        $.is(e.currentTarget as Element, 'input, textarea, button, select') &&
        !$.is(e.currentTarget as Element, 'input[type=range]')
      ) {
        return true
      }

      e.preventDefault()
      playback.toggle()

      return false
    },

    /**
     * Play the previous song when user presses K.
     */
    playPrev: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        $.is(e.currentTarget as Element, 'input, select, textarea')
      ) {
        return true
      }

      e.preventDefault()
      playback.playPrev()

      return false
    },

    /**
     * Play the next song when user presses J.
     */
    playNext: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        $.is(e.currentTarget as Element, 'input, select, textarea')
      ) {
        return true
      }

      e.preventDefault()
      playback.playNext()

      return false
    },

    /**
     * Put focus into the search field when user presses F.
     */
    search: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        ($.is(e.currentTarget as Element, 'input, select, textarea') &&
        !$.is(e.currentTarget as Element, 'input[type=range]'))
      ) {
        return true
      }

      if (e.metaKey || e.ctrlKey) {
        return true
      }

      e.preventDefault()
      event.emit(event.$names.FOCUS_SEARCH_FIELD)

      return false
    },

    /**
     * Like/unlike the current song when use presses L.
     */
    toggleLike: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        $.is(e.currentTarget as Element, 'input, select, textarea')
      ) {
        return true
      }

      if (!queueStore.current) {
        return false
      }

      favoriteStore.toggleOne(queueStore.current)
      socket.broadcast(event.$names.SOCKET_SONG, songStore.generateDataToBroadcast(queueStore.current))
      return false
    }
  },

  created: (): void => {
    if (KOEL_ENV === 'app') {
      listenToGlobalShortcuts()
    }
  }
})
</script>
