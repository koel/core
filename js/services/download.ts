import { playlistStore, favoriteStore } from '@/stores'
import { ls } from '.'
import { alerts } from '@/utils'

interface Download {
  fromSongs(songs: Song | Song[]): void
  fromAlbum(album: Album): void
  fromArtist(artist: Artist): void
  fromPlaylist(playlist: Playlist): void
  fromFavorites(): void
  trigger(uri: string): void
}

let events: any

if (KOEL_ENV === 'app') {
  events = require('&/events').default
}

export const download: Download = {
  fromSongs (songs: Song | Song[]): void {
    const query = (<Song[]>[]).concat(songs).reduce((q, song) => `songs[]=${song.id}&${q}`, '')
    this.trigger(`songs?${query}`)
  },

  fromAlbum (album: Album): void {
    this.trigger(`album/${album.id}`)
  },

  fromArtist (artist: Artist): void {
    this.trigger(`artist/${artist.id}`)
  },

  fromPlaylist (playlist: Playlist): void {
    if (playlistStore.getSongs(playlist).length) {
      this.trigger(`playlist/${playlist.id}`)
    }
  },

  fromFavorites (): void {
    if (favoriteStore.all.length) {
      this.trigger('favorites')
    }
  },

  /**
   * Build a download link using a segment and trigger it.
   *
   * @param  {string} uri The uri segment, corresponding to the song(s),
   *                      artist, playlist, or album.
   */
  trigger: (uri: string) => {
    const sep = uri.includes('?') ? '&' : '?'
    const url = `${window.BASE_URL}api/download/${uri}${sep}jwt-token=${ls.get('jwt-token')}`

    if (KOEL_ENV === 'app') {
      require('electron').ipcRenderer.send(events.DOWNLOAD, url)
      alerts.success('Download started!')
    } else {
      const iframe = document.createElement('iframe')
      iframe.style.display = 'none'
      iframe.setAttribute('src', url)
      document.body.appendChild(iframe)
    }
  }
}
