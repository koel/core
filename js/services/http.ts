import Axios, { AxiosInstance, Method } from 'axios'
import NProgress from 'nprogress'

import { event } from '@/utils/index'
import { ls } from '@/services'

export const http = {
  client: null as AxiosInstance | null,

  request (method: Method, url: string, data: object, successCb: any, errorCb: any): void {
    this.client!.request({ url, data, method }).then(successCb).catch(errorCb)
  },

  get (url: string, successCb?: any, errorCb?: any): void {
    this.request('get', url, {}, successCb, errorCb)
  },

  post (url: string, data: object, successCb?: any, errorCb?: any): void {
    this.request('post', url, data, successCb, errorCb)
  },

  put (url: string, data: object, successCb?: any, errorCb?: any): void {
    this.request('put', url, data, successCb, errorCb)
  },

  delete (url: string, data: object = {}, successCb?: any, errorCb?: any): void {
    this.request('delete', url, data, successCb, errorCb)
  },

  init (): void {
    this.client = Axios.create({
      baseURL: KOEL_ENV === 'app' ? `${ls.get('koelHost')}api` : `${window.BASE_URL}api`
    })

    // Intercept the request to make sure the token is injected into the header.
    this.client.interceptors.request.use(config => {
      config.headers.Authorization = `Bearer ${ls.get('jwt-token')}`
      return config
    })

    // Intercept the response and…
    this.client.interceptors.response.use(response => {
      // …get the token from the header or response data if exists, and save it.
      const token = response.headers.Authorization || response.data.token

      if (token) {
        ls.set('jwt-token', token)
      }

      NProgress.done()

      return response
    }, error => {
      // Also, if we receive a Bad Request / Unauthorized error
      if (error.response.status === 400 || error.response.status === 401) {
        // and we're not trying to login
        if (!(error.config.method === 'post' && /\/api\/me\/?$/.test(error.config.url))) {
          // the token must have expired. Log out.
          event.emit(event.$names.LOG_OUT)
        }
      }

      NProgress.done()

      return Promise.reject(error)
    })
  }
}
