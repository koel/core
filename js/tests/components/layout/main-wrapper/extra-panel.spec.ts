import each from 'jest-each'
import Component from '@/components/layout/main-wrapper/extra-panel.vue'
import factory from '@/tests/factory'
import { event } from '@/utils'
import { songInfo } from '@/services'
import { mock } from '@/tests/__helpers__'
import { shallow, Wrapper } from '@/tests/adapter'

const shallowComponent = (data: object = {}): Wrapper => shallow(Component, {
  stubs: ['lyrics-pane', 'artist-info', 'album-info', 'you-tube-video-list'],
  data: () => data
})

describe('components/layout/extra-panel', () => {
  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('renders properly', () => {
    expect(shallowComponent()).toMatchSnapshot()
  })

  it('does not have a YouTube tab if not using YouTube', async () => {
    const wrapper = shallowComponent({
      sharedState: {
        useYouTube: false
      }
    })
    await wrapper.vm.$nextTick()
    expect(shallow(Component)).toMatchSnapshot()
  })

  it('has a YouTube tab if using YouTube', async () => {
    const wrapper = shallowComponent({
      sharedState: {
        useYouTube: true
      }
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.has('you-tube-video-list-stub')).toBe(true)
  })

  each([['#extraTabLyrics'], ['#extraTabAlbum'], ['#extraTabArtist']])
    .test('switches to "%s" tab', selector => {
      expect(shallowComponent().click(selector).find('[aria-selected=true]').is(selector)).toBe(true)
    })

  it('fetch song info when a new song is played', () => {
    shallowComponent()
    const song = factory('song')
    const m = mock(songInfo, 'fetch', song)
    event.emit('SONG_PLAYED', song)
    expect(m).toHaveBeenCalledWith(song)
  })
})
