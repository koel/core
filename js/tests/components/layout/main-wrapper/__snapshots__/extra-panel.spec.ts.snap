// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/layout/extra-panel does not have a YouTube tab if not using YouTube 1`] = `
<section id="extra" class="showing">
  <div class="tabs">
    <div role="tablist" class="clear"><button aria-selected="true" aria-controls="extraPanelLyrics" id="extraTabLyrics" role="tab">
        Lyrics
      </button> <button aria-controls="extraPanelArtist" id="extraTabArtist" role="tab">
        Artist
      </button> <button aria-controls="extraPanelAlbum" id="extraTabAlbum" role="tab">
        Album
      </button>
      <!---->
    </div>
    <div class="panes">
      <div aria-labelledby="extraTabLyrics" id="extraPanelLyrics" role="tabpanel" tabindex="0">
        <lyricspane-stub song="[object Object]"></lyricspane-stub>
      </div>
      <div aria-labelledby="extraTabArtist" id="extraPanelArtist" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
      <div aria-labelledby="extraTabAlbum" id="extraPanelAlbum" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
      <div aria-labelledby="extraTabAlbum" id="extraPanelAlbum" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
    </div>
  </div>
</section>
`;

exports[`components/layout/extra-panel renders properly 1`] = `
<section id="extra" class="showing">
  <div class="tabs">
    <div role="tablist" class="clear"><button aria-selected="true" aria-controls="extraPanelLyrics" id="extraTabLyrics" role="tab">
        Lyrics
      </button> <button aria-controls="extraPanelArtist" id="extraTabArtist" role="tab">
        Artist
      </button> <button aria-controls="extraPanelAlbum" id="extraTabAlbum" role="tab">
        Album
      </button>
      <!---->
    </div>
    <div class="panes">
      <div aria-labelledby="extraTabLyrics" id="extraPanelLyrics" role="tabpanel" tabindex="0">
        <lyrics-pane-stub song="[object Object]"></lyrics-pane-stub>
      </div>
      <div aria-labelledby="extraTabArtist" id="extraPanelArtist" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
      <div aria-labelledby="extraTabAlbum" id="extraPanelAlbum" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
      <div aria-labelledby="extraTabAlbum" id="extraPanelAlbum" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
    </div>
  </div>
</section>
`;
